### PR TITLE
Custom dialogs DialogSettings

### LINKED ISSUE #2735: MetroWindow.ShowMetroDialogAsync odd behaviour with provided DialogSettings
## What steps will reproduce this issue?
Calling
```
var myDialogSettings = new MetroDialogSettings(){AnimateHide = false,AnimateShow = false};
metroWindow.ShowMetroDialogAsync(myDialog, myDialogSettings)
```

does not use the provided DialogSettings **myDialogSettings**. Instead the DialogSettings of the base class of **myDialog** are used (and these are by default AnimateHide = AnimateShow = true)

### Expected outcome

I think the provided DialogSettings should be used. More over I think that the provided DialogSettings **myDialogSettings** should be set on the dialog instance **myDialog** because normally there is a close handler within **myDialog** that calls

`metroWindow.HideMetroDialogAsync(this, this.DialogSettings);`.

That way ShowMessageAsync(..) and ShowMetroDialogAsync(..) would behave equally.

See example:
[ShowMetroDialogAsyncIssue.zip](https://github.com/MahApps/MahApps.Metro/files/571266/ShowMetroDialogAsyncIssue.zip)

--
### Environment

- MahApps.Metro __v1.3.0__
- Windows __7x64__
- Visual Studio __2013__
- .NET Framework __4.5__

### PR DESCRIPTION
**Pull request by punker76**
## What changed?

Some changes to fix #2735 MetroWindow.ShowMetroDialogAsync odd behaviour with provided DialogSettings

Closes #2735

### CODE REVIEW

File: src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/DialogManager.cs [hunk 1]
```diff
@@ -375,7 +406,8 @@ public static Task HideMetroDialogAsync(this MetroWindow window, BaseMetroDialog
                 {
                     window.RemoveDialog(dialog);
 
-                    return HandleOverlayOnHide(settings,window);
+                    settings = settings ?? dialog.DialogSettings;
```
Comment (thoemmi) [inline src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/DialogManager.cs]: Is this ok, or should it be `settings = settings ?? (dialog.DialogSettings ?? window.MetroDialogOptions);`

**Comment by punker76** [comment]
## It's now possible to do...

- `var customBaseMetroDialog = await this.ShowMetroDialogAsync<CustomBaseMetroDialog>();`
creates and shows a new custom dialog of the given type with the default dialog settings
- `var customBaseMetroDialog = await this.ShowMetroDialogAsync<CustomBaseMetroDialog>(this.metroDialogSettings);`
creates and shows also a new custom dialog of the given type with own settings

the custom dialog for this samples
```xaml
<dialogs:CustomDialog x:Class="ShowMetroDialogAsyncIssue.CustomBaseMetroDialog"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      xmlns:dialogs="clr-namespace:MahApps.Metro.Controls.Dialogs;assembly=MahApps.Metro"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008" 
      mc:Ignorable="d" 
      d:DesignHeight="300" d:DesignWidth="300"
      Title="CustomBaseMetroDialog">
    <StackPanel>
        <TextBlock Text="This is a custom dialog based on BaseMetroDialog" />
        <Button Click="Close_OnClick">Close</Button>
    </StackPanel>
</dialogs:CustomDialog>
```

```csharp
namespace ShowMetroDialogAsyncIssue
{
    using System.Windows;
    using MahApps.Metro.Controls;
    using MahApps.Metro.Controls.Dialogs;

    public partial class CustomBaseMetroDialog : CustomDialog
    {
        public CustomBaseMetroDialog()
        {
            InitializeComponent();
        }

        public CustomBaseMetroDialog(MetroDialogSettings settings) : base(settings)
        {
            InitializeComponent();
        }

        public CustomBaseMetroDialog(MetroWindow parentWindow, MetroDialogSettings settings) : base(parentWindow, settings)
        {
            InitializeComponent();
        }

        private void Close_OnClick(object sender, RoutedEventArgs e)
        {
            (this.OwningWindow ?? (MetroWindow)Application.Current.MainWindow).HideMetroDialogAsync(this);
        }
    }
}
```

## Or
- `await this.ShowMetroDialogAsync(new CustomBaseMetroDialog());
create and show custom dialog with default settings
- `await this.ShowMetroDialogAsync(new CustomBaseMetroDialog(this.metroDialogSettings));` or `await this.ShowMetroDialogAsync(new CustomBaseMetroDialog(this, this.metroDialogSettings));`
create and show custom dialog with own settings
## And..
It's possible to move the settings stuff to the custom dialog itself.
```csharp
protected override MetroDialogSettings OnGetDialogSettings(MetroDialogSettings settings)
{
    return new MyCustomDialogSettings();
}
```

/cc @thoemmi

File: src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/BaseMetroDialog.cs [hunk 2]
```diff
@@ -88,25 +89,32 @@ static BaseMetroDialog()
...
-            this.DialogSettings = new MetroDialogSettings();
+        }
 
-            this.Initialize();
+        /// <summary>
+        /// With this method it's possible to return your own settings in a custom dialog.
+        /// </summary>
+        /// <param name="settings"></param>
+        /// <returns></returns>
+        protected virtual MetroDialogSettings OnGetDialogSettings(MetroDialogSettings settings)
```
Comment (thoemmi) [inline src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/BaseMetroDialog.cs]: I would call this method `ConfigureSettings`. E.g. in ASP.NET you have methods like `ConfigureServices`, which allow you to intercept and add additional services. Here it allows the user to intercept and change the settings object.

Comment (punker76) [inline src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/BaseMetroDialog.cs]: Thx @thoemmi, naming is not my favorite ;-)

Comment (thoemmi) [inline src/MahApps.Metro/MahApps.Metro.Shared/Controls/Dialogs/BaseMetroDialog.cs]: There two hard problems in computer science:
* cache invalidation,
* naming things, and
* off-by-1 errors

😉

**Pull request merged**